src/GraphControl/DirectX/RenderMain.cpp

```diff
@@ -210,53 +201,37 @@ namespace GraphControl::DX
 
     bool RenderMain::RunRenderPass()
     {
-        // Non async render passes cancel if they can't obtain the lock immediatly
-        if (!m_criticalSection.try_lock())
-        {
-            return false;
-        }
-
-        m_criticalSection.unlock();
-
-        critical_section::scoped_lock lock(m_criticalSection);
-
         return RunRenderPassInternal();
     }
 
-    IAsyncAction ^ RenderMain::RunRenderPassAsync(bool allowCancel)
+    concurrency::task<bool> RenderMain::RunRenderPassAsync(bool allowCancel)
     {
-        // Try to cancel the renderPass that is in progress
-        if (m_renderPass != nullptr && m_renderPass->Status == ::AsyncStatus::Started)
+        if (allowCancel)
```

Comment:
Just according to the original implementation of this method, originally `allowCancel` is for the task being fired. For example, task 1 is fired with `false`, then when firing task 2, no matter `allowCancel` is true or false, the task 1 will not be canceled.
But after this change, task 2 with `true` will cancel the task 1. Not sure if this is as expected.

Reply:
Good point. Yes, the cancellation point was not doing right. I updated it to the first place where the task gets resumed.
Also, I removed the cancellation token and choose to use an integer to tag a version for each async task. Because it is much cheaper and works well in the same thread.